### ISSUE TITLE
Better Query Generation For Left Join Statements With Filters On The Right Hand Side

### ISSUE THREAD

**Issue by brantburnett** [couchbaselabs/Linq2Couchbase#218]
This is a continuation of issue #199.  The issue is that the DocumentTypeFilter always applies a predicate to the bucket, even if the bucket is being referenced in a JOIN statement with ON KEYS.  In traditional SQL, such a predicate would be applied in the "ON" clause of the join, but that isn't an option in N1QL.

For INNER JOIN statements, this predicate can be safely hoisted to the main WHERE clause:

```
SELECT * FROM `beer-sample` beer
INNER JOIN `beer-sample` brewery ON KEYS beer.brewery_id
WHERE beer.type = 'beer' AND brewery.type = 'brewery'
```

However, hoisting is not safe for LEFT JOIN statements:

```
SELECT * FROM `beer-sample` beer
LEFT JOIN `beer-sample` brewery ON KEYS beer.brewery_id
WHERE beer.type = 'beer' AND brewery.type = 'brewery'
```

If the filter on `brewery.type` is applied *before* the join (as a DocumentTypeFilter does), then this query incorrectly filters any beers where the brewery is not found, or where the only breweries found have type <> 'brewery'.  The expected behavior for the LINQ syntax would be to retain any such beers with no values for the brewery.

This has been previously addressed with a workaround, where you can suppress the DocumentTypeFilter on the right-hand side of the join (see pull #200).  However, this is just a workaround.  Ideally, it should not be required.  It also doesn't assist with scenarios where the user does want to apply filters prior to the join.

This syntax may be a workable alternative (so far appears to work in 4.6.3 and 5.0.0).  However, further testing is required for compatibility and to ensure there are no edge cases where it does not function:

```
SELECT *
FROM `beer-sample` as `Extent1`
LEFT UNNEST (SELECT `Extent3`.* FROM `beer-sample` as `Extent3` USE KEYS `Extent1`.`brewery_id` WHERE `Extent3`.`type` = 'brewery') as `Extent2`
WHERE (`Extent1`.`type` = 'beer')
```

**Comment by jeffrymorris** [comment]
Thanks @brantburnett for the in-depth description.

**Comment by brantburnett** [comment]
Upon further research, the approach I suggested seems to have a poor performance profile.  I ran tests against beer-sample joining beers to breweries, and it was approximately 5 times slower than a plain LEFT JOIN on Couchbase Server 5.0.0.

Based on [this conversation](https://forums.couchbase.com/t/left-outer-join-with-filters-on-the-right-side-extent/14513) it appears a more solid solution will be built into N1QL, potentially as soon as 5.1 🤞.  Based on this, I think we're better off to wait until then and implement the clean solution instead of a workaround.
